src/kpro_rsp_lib.erl

```diff
@@ -0,0 +1,209 @@
+%%%   Copyright (c) 2014-2018, Klarna AB
+%%%
+%%%   Licensed under the Apache License, Version 2.0 (the "License");
+%%%   you may not use this file except in compliance with the License.
+%%%   You may obtain a copy of the License at
+%%%
+%%%       http://www.apache.org/licenses/LICENSE-2.0
+%%%
+%%%   Unless required by applicable law or agreed to in writing, software
+%%%   distributed under the License is distributed on an "AS IS" BASIS,
+%%%   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%%%   See the License for the specific language governing permissions and
+%%%   limitations under the License.
+%%%
+
+-module(kpro_rsp_lib).
+
+-export([ decode/4
+        , dec_struct/4
+        ]).
+
+-export([ parse/1
+        ]).
+
+-include("kpro_private.hrl").
+
+-define(IS_STRUCT_SCHEMA(Schema), is_list(Schema)).
+
+%% @doc Decode message body binary (without the leading 4-byte correlation ID.
+-spec decode(kpro:api(), kpro:vsn(), binary(), reference()) -> kpro:rsp().
+decode(API, Vsn, Body, Ref) ->
+  {Message, <<>>} =
+    try
+      decode_struct(API, Vsn, Body)
+    catch error : E ->
+      Context = [ {api, API}
+                , {vsn, Vsn}
+                , {body, Body}
+                ],
+      Trace = erlang:get_stacktrace(),
+      erlang:raise(error, {E, Context}, Trace)
+    end,
+  #kpro_rsp{ ref = Ref
+           , api = API
+           , vsn = Vsn
+           , msg = Message
+           }.
+
+-spec parse(kpro:rsp()) -> term().
+parse(#kpro_rsp{ api = list_offsets
+               , msg = Msg
+               }) ->
+  case get_partition_rsp(Msg) of
+    #{offsets := [Offset]} = M -> M#{offset => Offset};
+    #{offset := _} = M -> M
+  end;
+parse(#kpro_rsp{ api = produce
+               , msg = Msg
+               }) ->
+  get_partition_rsp(Msg);
+parse(#kpro_rsp{ api = fetch
+               , vsn = Vsn
+               , msg = Msg
+               }) ->
+  ErrorCode = kpro:find(error_code, Msg, ?kpro_no_error),
+  SessionID = kpro:find(session_id, Msg, 0),
+  {Header, Batches} =
+    case kpro:find(responses, Msg) of
+      [] ->
+        %% a session init without data
+        {undefined, []};
+      _ ->
+        PartitionRsp = get_partition_rsp(Msg),
+        Header0 = kpro:find(partition_header, PartitionRsp),
+        Records = kpro:find(record_set, PartitionRsp),
+        {Header0, decode_batches(Vsn, Records)}
+    end,
+  #{ error_code => ErrorCode
+   , session_id => SessionID
+   , header => Header
+   , batches => Batches
+   };
+parse(#kpro_rsp{ api = create_topics
+               , msg = Msg
+               }) ->
+  Errors = kpro:find(topic_errors, Msg),
+  Pred = fun(#{error_code := EC}) -> EC =/= ?kpro_no_error end,
```

Comment:
You're repeating this anonymous function in several places.  Maybe add a named function for it?